src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -312,59 +340,34 @@ Visitor::visitTypesafe(Parser::GroupGraphPatternSubContext* ctx) {
-  auto [ops, filters] =
-      splitUp(visitVector<std::variant<GraphPatternOperation, SparqlFilter>>(
-          ctx->graphPatternNotTriples()));
-  ops.emplace_back(
-      GraphPatternOperation::BasicGraphPattern{std::move(triples)});
-  return {ops, filters};
-}
-
-// ____________________________________________________________________________________
-vector<TripleWithPropertyPath> Visitor::visitTypesafe(
-    Parser::TriplesBlockContext* ctx) {
-  auto triples = visitTypesafe(ctx->triplesSameSubjectPath());
+  BasicGraphPattern triples = {ad_utility::transform(
+      visitTypesafe(ctx->triplesSameSubjectPath()), convertAndRegisterTriple)};
   if (ctx->triplesBlock()) {
-    ad_utility::appendVector(triples, visitTypesafe(ctx->triplesBlock()));
+    triples.appendTriples(visitTypesafe(ctx->triplesBlock()));
   }
-  return triples;
+  return {triples};
 }
 
 // ____________________________________________________________________________________
-std::variant<GraphPatternOperation, SparqlFilter> Visitor::visitTypesafe(
+Visitor::OperationOrFilter Visitor::visitTypesafe(
     Parser::GraphPatternNotTriplesContext* ctx) {
-  // TODO: correctly set optional and id attributes
   if (ctx->graphGraphPattern() || ctx->serviceGraphPattern()) {
-    throw ParseException(
-        "GraphGraphPattern or ServiceGraphPattern are not supported.");
-  } else if (ctx->filterR()) {
-    // TODO: visitAlternative
-    return visitTypesafe(ctx->filterR());
-  } else if (ctx->optionalGraphPattern()) {
-    auto optionalPattern = visitTypesafe(ctx->optionalGraphPattern());
-    optionalPattern._child._optional = true;
-    return GraphPatternOperation{std::move(optionalPattern)};
-  } else if (ctx->minusGraphPattern()) {
-    return GraphPatternOperation{visitTypesafe(ctx->minusGraphPattern())};
-  } else if (ctx->bind()) {
-    return GraphPatternOperation{visitTypesafe(ctx->bind())};
-  } else if (ctx->inlineData()) {
-    return GraphPatternOperation{visitTypesafe(ctx->inlineData())};
-  } else if (ctx->groupOrUnionGraphPattern()) {
-    return visitTypesafe(ctx->groupOrUnionGraphPattern());
+    reportError(ctx,
+                "GraphGraphPattern or ServiceGraphPattern are not supported.");
   } else {
-    AD_FAIL();  // Unreachable.
+    return visitAlternative<std::variant<GraphPatternOperation, SparqlFilter>>(
+        ctx->filterR(), ctx->optionalGraphPattern(), ctx->minusGraphPattern(),
+        ctx->bind(), ctx->inlineData(), ctx->groupOrUnionGraphPattern());
   }
 }
 
 // ____________________________________________________________________________________
-GraphPatternOperation::Optional Visitor::visitTypesafe(
+GraphPatternOperation Visitor::visitTypesafe(
     Parser::OptionalGraphPatternContext* ctx) {
-  return {visitTypesafe(ctx->groupGraphPattern())};
+  auto pattern = visitTypesafe(ctx->groupGraphPattern());
+  pattern._optional = true;
+  return GraphPatternOperation{
+      GraphPatternOperation::Optional{std::move(pattern)}};
```

Comment:
I think I have found the optimal solution for now:
Make the constructor of `Optional` enforce the invariant and set the `_optional = true`. 
There we can easily refactor it, and that's what constructors are for.